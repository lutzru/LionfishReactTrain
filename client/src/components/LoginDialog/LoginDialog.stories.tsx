import { action } from '@storybook/addon-actions'
import { text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import 'utils/ i18n'
import LoginDialog, { LoginDialogProps } from './LoginDialog'

const stories = storiesOf('LoginDialog', module)

const props: LoginDialogProps = {
    userName: 'TestUsername',
    password: 'geheim',
    loginError: 'Du kommst hier ned rein.',
    onClick: action('onClick'),
    onUserNameChange: action('onUserNameChange'),
    onPasswordChange: action('onPasswordChange'),
}

const props2: LoginDialogProps = {
    userName: '',
    password: '',
    onClick: action('onClick'),
    onUserNameChange: action('onUserNameChange'),
    onPasswordChange: action('onPasswordChange'),
}

const props3: LoginDialogProps = {
    userName: 'Lutz',
    password: '',
    onClick: action('onClick'),
    onUserNameChange: action('onUserNameChange'),
    onPasswordChange: action('onPasswordChange'),
}

const props4: LoginDialogProps = {
    userName: 'Lutz',
    password: 'geheim',
    onClick: action('onClick'),
    onUserNameChange: action('onUserNameChange'),
    onPasswordChange: action('onPasswordChange'),
}

stories.add('empty view', () => <LoginDialog {...props2} />)
stories.add('only Username', () => <LoginDialog {...props3} />)
stories.add('Username and password given', () => <LoginDialog {...props4} />)
stories.add('view with error', () => <LoginDialog {...props} />)

//
// Bei einem Interactive view:
// Werte mÃ¼ssen hier drin definiert sein,
// nicht als const vorher. (Damit nobs funktioniert.)
//

stories.addDecorator(withKnobs)

stories.add('interactive view', () => (
    <LoginDialog
        userName={text('Username', '')}
        password={text('Passwort', '')}
        loginError={text('ErrorMessage', '')}
        onClick={action('onClick')}
        onUserNameChange={action('onUserNameChange')}
        onPasswordChange={action('onPasswordChange')}
    />
))
