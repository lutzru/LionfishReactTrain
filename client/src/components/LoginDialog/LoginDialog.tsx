import { Box, Button, Grid, Paper, TextField, Typography } from '@material-ui/core'
import { Alert, AlertTitle } from '@material-ui/lab'
import * as React from 'react'

export interface LoginDialogProps {
    userName: string
    password: string
    loginError?: string // ? - optionales Feld
    onClick: () => void
    onUserNameChange: (userName: string) => void
    onPasswordChange: (password: string) => void
}

const LoginDialog = (props: LoginDialogProps): JSX.Element => {
    return (
        <Box padding={3}>
            <Paper elevation={3}>
                <Box padding={3}>
                    <Grid container direction="column" spacing={3}>
                        <Grid item>
                            <Typography variant="h2">Login:</Typography>
                        </Grid>
                        {props.loginError && (
                            <Grid item>
                                <Alert severity="error">
                                    <AlertTitle>Error</AlertTitle>
                                    {props.loginError}
                                </Alert>
                            </Grid>
                        )}
                        <Grid item>
                            <TextField
                                id="standard-basic"
                                label="Benutzername"
                                value={props.userName}
                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                                    props.onUserNameChange(e.target.value)
                                }}
                            />
                        </Grid>
                        <Grid item>
                            <TextField
                                id="standard-basic"
                                label="Passwort"
                                type="password"
                                value={props.password}
                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                                    props.onPasswordChange(e.target.value)
                                }}
                            />
                        </Grid>

                        <Grid item>
                            <Button
                                variant="contained"
                                color="primary"
                                onClick={props.onClick}
                                disabled={!props.userName || !props.password}
                            >
                                Go
                            </Button>
                        </Grid>
                    </Grid>
                </Box>
            </Paper>
        </Box>
    )
}

export default LoginDialog
