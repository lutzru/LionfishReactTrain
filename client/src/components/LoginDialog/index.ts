// INFO: index.ts, um es schön kurz importieren zu können

import LoginDialog from './LoginDialog'

export default LoginDialog
