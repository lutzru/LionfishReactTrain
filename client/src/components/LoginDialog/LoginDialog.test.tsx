import { Typography } from '@material-ui/core'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import LoginDialog, { LoginDialogProps } from './LoginDialog'

jest.mock('react-i18next', () => ({
    useTranslation: (): unknown => ({ t: (key: string): string => key }),
}))
jest.mock('react', () => ({
    // @ts-ignore
    ...jest.requireActual('react'),
    useState: jest.fn(),
}))

const props: LoginDialogProps = {
    userName: 'TestUsername',
    password: 'geheim',
    loginError: 'Du kommst hier ned rein.',
    onClick: jest.fn(),
    onUserNameChange: jest.fn(),
    onPasswordChange: jest.fn(),
}

describe('LoginDialog Component', () => {
    beforeEach(() => {
        const setState = jest.fn()
        // @ts-ignore
        useStateMock.mockImplementation((state) => [state, setState])
    })
    it('should render correctly with react-test-renderer', () => {
        const component = renderer.create(<LoginDialog {...props} />)

        expect(component).toMatchSnapshot()
    })
    it('should render correctly with enzyme', () => {
        const component = shallow(<LoginDialog {...props} />)

        expect(toJson(component)).toMatchSnapshot()
    })
    it('should handle a click correctly', () => {
        const onClickMock = jest.fn()
        const component = shallow(<LoginDialog {...props} />)

        component.find(Typography).simulate('click', { currentTarget: { value: 0 } })
        expect(onClickMock).toHaveBeenCalledTimes(1)
    })
})
