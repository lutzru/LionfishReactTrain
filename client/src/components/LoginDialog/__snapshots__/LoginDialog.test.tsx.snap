// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDialog Component should render correctly with enzyme 1`] = `
<Styled(MuiBox)
  padding={3}
>
  <WithStyles(ForwardRef(Paper))
    elevation={3}
  >
    <Styled(MuiBox)
      padding={3}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        direction="column"
        spacing={3}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Typography))
            variant="h2"
          >
            Login:
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Alert))
            severity="error"
          >
            <WithStyles(ForwardRef(AlertTitle))>
              Error
            </WithStyles(ForwardRef(AlertTitle))>
            Du kommst hier ned rein.
          </WithStyles(ForwardRef(Alert))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            id="standard-basic"
            label="Benutzername"
            onChange={[Function]}
            value="TestUsername"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            id="standard-basic"
            label="Passwort"
            onChange={[Function]}
            type="password"
            value="geheim"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disabled={false}
            onClick={[MockFunction]}
            variant="contained"
          >
            Go
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Paper))>
</Styled(MuiBox)>
`;
